src/modules/partition/core/PartitionLayout.cpp

```diff
@@ -164,60 +164,118 @@ PartitionLayout::execute( Device* dev,
     qint64 minSize, maxSize, end;
     qint64 totalSize = lastSector - firstSector + 1;
     qint64 availableSize = totalSize;
+    qint64 minPartitionSize = 0;
 
-    // TODO: Refine partition sizes to make sure there is room for every partition
-    // Use a default (200-500M ?) minimum size for partition without minSize
-
+    // Let's check if we have enough space for each partSize
     foreach ( const PartitionLayout::PartitionEntry& part, m_partLayout )
     {
-        Partition* currentPartition = nullptr;
-
         qint64 size = -1;
         // Calculate partition size
+
         if ( part.partSize.isValid() )
         {
-            size = part.partSize.toSectors( totalSize, dev->logicalSize() );
+            // We need to ignore the percent-defined
+            if ( part.partSize.unit() != CalamaresUtils::Partition::SizeUnit::Percent)
+            {
+                size = part.partSize.toSectors( totalSize, dev->logicalSize() );
+            }
+            else
+            {
+                if ( part.partMinSize.isValid() )
+                {
+                    size = part.partMinSize.toSectors( totalSize, dev->logicalSize() );
+                }
+                else
+                {
+                    size = 0;
+                }
+            }
         }
         else
         {
             cWarning() << "Partition" << part.partMountPoint << "size (" << size << "sectors) is invalid, skipping...";
             continue;
         }
 
-        if ( part.partMinSize.isValid() )
-        {
-            minSize = part.partMinSize.toSectors( totalSize, dev->logicalSize() );
-        }
-        else
-        {
-            minSize = 0;
-        }
+        partSizeMap.insert (&part, size);
+        availableSize -= size;
+    }
 
-        if ( part.partMaxSize.isValid() )
-        {
-            maxSize = part.partMaxSize.toSectors( totalSize, dev->logicalSize() );
-        }
-        else
+    // Use partMinSize and see if we can do better afterward.
+    if (availableSize < 0)
+    {
+        availableSize = totalSize;
+        foreach ( const PartitionLayout::PartitionEntry& part, m_partLayout )
         {
-            maxSize = availableSize;
+            qint64 size;
+
+            if ( part.partMinSize.isValid() )
+            {
+                size = part.partMinSize.toSectors( totalSize, dev->logicalSize() );
+            }
+            else if ( part.partSize.isValid() )
+            {
+                if ( part.partSize.unit() != CalamaresUtils::Partition::SizeUnit::Percent)
+                {
+                    size = part.partSize.toSectors( totalSize, dev->logicalSize() );
+                }
+                else
+                {
+                    size = 0;
+                }
+            }
+            else
+            {
+                size = 0;
+            }
+
+            partSizeMap.insert (&part, size);
+            availableSize -= size;
         }
+    }
```

Comment:
So this loop is the same as the previous, only preferring min-size over size (and the previous was preferring size over min-size). That strikes me as a candidate for refactoring with some kind of lambda to get preferred- and secondary-sizes. Again, that might be more code-churn than you want to do.